Branch-SDK-TestBed/src/io/branch/branchandroiddemo/test/BUOTestRoutines.java

```diff
@@ -0,0 +1,118 @@
+                                    .setProductBrand("test_prod_brand")
+                                    .setProductCategory(ProductCategory.APPAREL_AND_ACCESSORIES)
+                                    .setProductName("test_prod_name")
+                                    .setProductVariant("test_prod_variant")
+                                    .setQuantity(1.5)
+                                    .setSku("test_sku")
+                                    .setContentSchema(BranchContentSchema.COMMERCE_PRODUCT)
+
+                    )
+                    .addKeyWord("keyword1")
+                    .addKeyWord("keyword2");
+            succeeded = doTestBUOSerialisation(testBuo);
+        }
+        if (succeeded) {
+            Log.d("BranchTestBed", "Passed BUO serialisation - de-serialisation test.");
+        } else {
+            Log.d("BranchTestBed", "Failed BUO serialisation - de-serialisation test.");
+        }
+        return succeeded;
+    }
+
+    private static boolean doTestBUOSerialisation(BranchUniversalObject buo) {
+        boolean isPassed = false;
+        JSONObject testBuoJson1 = buo.convertToJson();
+        BranchUniversalObject buo2 = BranchUniversalObject.createInstance(testBuoJson1);
+        JSONObject testBuoJson2 = buo2.convertToJson();
+
+        Parcel parcel = Parcel.obtain();
+        buo.writeToParcel(parcel, 0);
+        parcel.setDataPosition(0);
+        BranchUniversalObject buoCreatedFromParcel = (BranchUniversalObject) BranchUniversalObject.CREATOR.createFromParcel(parcel);
+
+        return (checkIfIdenticalJson(buo.getContentMetadata().convertToJson(), buo2.getContentMetadata().convertToJson()))
+                && (checkIfIdenticalJson(testBuoJson1, testBuoJson2))
+                && (checkIfIdenticalJson(testBuoJson1, buoCreatedFromParcel.convertToJson()));
+
+
+    }
+
+    private static boolean checkIfIdenticalJson(JSONObject obj1, JSONObject obj2) {
+        boolean isIdentical = false;
+        if (obj1.length() == obj2.length()) {
+            Iterator<String> keys = obj1.keys();
+            try {
+                while (keys.hasNext()) {
+                    String currKey = keys.next();
+                    if (obj1.get(currKey).equals(obj2.opt(currKey))) {
+                        obj2.remove(currKey);
+                    }
+                }
+                if (obj2.length() == 0) {
+                    isIdentical = true;
+                } else {
+                    Log.e("BranchTestBed", "Error : BUO serialisation error. Reason: Additional entries in de-serialised object " + obj2);
+                }
+            } catch (JSONException e) {
+                e.printStackTrace();
+                Log.e("BranchTestBed", "Error : BUO serialisation error.");
+            }
+        } else {
+            Log.e("BranchTestBed", "Serialised versions buo versions are not matching \n" + obj1 + "\n\n" + obj2);
+        }
+
+        return isIdentical;
+    }
+
+}
```

Comment:
love the testing! I didn't look to see if it covers every single field. Let me know if you want help with that.